core/dbt/adapters/base/impl.py

```diff
@@ -162,12 +175,62 @@ def submit(self, compiled_code: str) -> Any:
         raise NotImplementedError("PythonJobHelper submit function is not implemented yet")
 
 
-class AdapterFeature(str, Enum):
+class Capability(str, Enum):
     """Enumeration of optional adapter features which can be probed using BaseAdapter.has_feature()"""
 
-    CatalogByRelations = "CatalogByRelations"
-    """Flags support for retrieving catalog information using a list of relations, rather than always retrieving all
-    the relations in a schema """
+    SchemaMetadataByRelations = "CatalogByRelations"
+    """Indicates efficient support for retrieving schema metadata for a list of relations, rather than always retrieving
+    all the relations in a schema."""
+
+    TableLastModifiedMetadata = "TableLastModifiedMetadata"
+    """Indicates support for determining the time of the last table modification by querying database metadata."""
+
+
+class Support(str, Enum):
+    Unknown = "Unknown"
+    """The adapter has not declared whether this capability is a feature of the underlying DBMS."""
+
+    Unsupported = "Unsupported"
+    """This capability is not possible with the underlying DBMS, so the adapter does not implement related macros."""
+
+    NotImplemented = "NotImplemented"
+    """This capability is available in the underlying DBMS, but support has not yet been implemented in the adapter."""
+
+    Versioned = "Versioned"
+    """Some versions of the DBMS supported by the adapter support this capability and the adapter has implemented any
+    macros needed to use it."""
+
+    Full = "Full"
+    """All versions of the DBMS supported by the adapter support this capability and the adapter has implemented any
+    macros needed to use it."""
+
+
+@dataclasses.dataclass
+class CapabilitySupport:
+    capability: Capability
+    support: Support
+    first_version: Optional[str] = None
+
+    def __bool__(self):
+        return self.support == Support.Versioned or self.support == Support.Full
+
+
+# Make python 3.8 happy with this generic type, remove when we support 3.9+
+if TYPE_CHECKING:
+    CapabilityUserDict = UserDict[Capability, CapabilitySupport]
```

Comment:
I'm still not clear on why we need a `dict` to assign a `CapabilitySupport` to a `Capability` when an attribute of `CapabilitySupport` is its `Capability`. I get the value in being able to look up the support by capability, but perhaps we don't need `capability` on `CapabilitySupport` then?

Reply:
> perhaps we don't need capability on CapabilitySupport then?

Agree that it's not necessary in the current design, and I'll get rid of that.

Reply:
Done.